### DISCUSSION TITLE
Extend library

### DISCUSSION THREAD

**Discussion by xenonCBU** [angularsen/UnitsNet#1064] [Q&A]
Hello ! Do I have any chance to extend this library ? I want to implement IXmlSerializable in order to parse the units using XmlSerializer from Microsoft. It is enough for me to have something like 5V for 5 volts

**Comment by angularsen** [comment]
Hi, are you able to use [DataContractSerializer](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer?view=net-6.0) ?

```cs
MemoryStream ms = new MemoryStream();
DataContractSerializer ser =
	new DataContractSerializer(typeof(Length));
ser.WriteObject(ms, Length.FromCentimeters(100));
ms.Position = 0;
new StreamReader(ms).ReadToEnd().Dump();
ms.Close();
```

```xml
<Length xmlns="http://schemas.datacontract.org/2004/07/UnitsNet" xmlns:i="http://www.w3.org/2001/XMLSchema-instance">
    <Value>100</Value>
    <Unit>Centimeter</Unit>
</Length>
```

**Reply by xenonCBU** [reply]
The ideea is that I want to use [XMLSerilizer](https://docs.microsoft.com/en-us/dotnet/api/system.xml.serialization.ixmlserializable?view=net-6.0) like that 
ElectricPotential electric = ElectricPotential.FromVolts(5);
XmlSerilizers xs = new XmlSerializer(typeof(ElectricPotential));
TextWriter txt = new StreamWriter(@path@);
xs.Serialize(txt, electric);

I can implement express a method for serialization, but I want to add an instance of ElectricPotential in another class and then to serialize the entire class and I need a standard xml parser like that one from microsoft which know how to parse everything.

**Reply by angularsen** [reply]
Sorry, but I don't think we will add `IXmlSerializable` to all our quantities.

We already support XML via [DataContractSerializer](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer?view=net-6.0).

I suggest you wrap the quantities in your own type that implements IXmlSerializable instead, then you read/write those, and parse back into UnitsNet quantities. Then you are more in control of any future breaking changes of UnitsNet too.

**Reply by xenonCBU** [reply]
How can I wrap it  ? 
I tried with inheritance but the units are sealed, then I tried to create an extension, but if I want to define the methods of IXMLSerializer (ReadXml and WriteXml) then I have to add as parameter something like this : public static void ReadXml(this ElectricPotential objElectricPotential, XmlReader reader) {...} and the serializer doesn't recognize the method .

**Reply by angularsen** [reply]
You can create your own class with `double Value`, `string Unit`, `string QuantityType` properties, then serialize/deserialize that to XML. Then you are in control of what values are serialized, and how to translate between your data storage types and UnitsNet types.
